### PR TITLE
[IMPROVEMENT] General stability and bug fixes.

### PR DESCRIPTION
**Pull request by derekcollison**
This PR has general improvements and fixes to filestore, raft, and the clustering layer.

Summary

1. Additional support for preAck handling for interest based streams when replicated acks arrive before the message itself.
2. Better handling when checking state to determine whether to remove an interest based message.
3. Improved StepDown() and leadership transfer handling after restarts.
4. Improved voting logic for high load systems.
5. Various improvements and fixes for filestore Compact(), which is used heavily in the raft layer when updating snapshots and the raft wal.

Signed-off-by: Derek Collison <derek@nats.io>

### CODE REVIEW

File: server/jetstream_cluster.go [hunk 1]
```diff
@@ -7316,25 +7386,39 @@ func (mset *stream) processSnapshot(snap *streamSnapshot) (e error) {
...
+		}
+		mset.mu.Unlock()
+
+		if didReset {
+			s.Warnf("Catchup for stream '%s > %s' resetting first sequence: %d on catchup complete",
+				mset.account(), mset.name(), firstExpected)
+		}
+
+		mset.mu.RLock()
+		var consumers []*consumer
```
Comment (neilalexander) [inline server/jetstream_cluster.go:7415]:
```suggestion
		consumers := make([]*consumer, 0, len(mset.consumers))
```

File: server/stream.go [hunk 2]
```diff
@@ -5076,37 +5110,17 @@ func (a *Account) RestoreStream(ncfg *StreamConfig, r io.Reader) (*stream, error
...
 }
 
-// This is to check for dangling messages.
+// This is to check for dangling messages on interest retention streams.
 // Issue https://github.com/nats-io/nats-server/issues/3612
 func (mset *stream) checkForOrphanMsgs() {
-	// We need to grab the low water mark for all consumers.
-	var ackFloor uint64
 	mset.mu.RLock()
+	var consumers []*consumer
```
Comment (neilalexander) [inline server/stream.go:5117]:
```suggestion
	consumers := make([]*consumer, 0, len(mset.consumers))
```

**Review by neilalexander — APPROVED** [review]
Overall looks fine to me, just a couple minor things!

File: server/filestore.go [hunk 3]
```diff
@@ -106,6 +106,7 @@ type psi struct {
 }
 
 type fileStore struct {
+	srv         *Server
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/filestore.go:109]: Mostly to allow access for logging or debugging.

**Review by wallyqs — APPROVED** [review]
LGTM

File: server/consumer.go [hunk 4]
```diff
@@ -2495,37 +2500,40 @@ func (o *consumer) needAck(sseq uint64, subj string) bool {
...
 			}
 			subj = svp.subj
 		}
 		if !o.isFilteredMatch(subj) {
-			o.mu.RUnlock()
 			return false
 		}
 	}
 
+	if isFiltered && o.mset == nil {
```
Comment (Jarema) [inline server/consumer.go:2522]: feels more natural to have this `mset == nil` check and return  before `mset != nil` check?

Comment (derekcollison) [reply server/consumer.go:2522]: Will rearrange.

**Pull request merged**